### PR TITLE
Implement `flock_*` fiber-aware, without blocking the thread

### LINKED ISSUE #12721: Fiber-aware blocking `flock`
`FileDescriptor#flock_exclusive(blocking: true)` and `FileDescriptor#flock_shared(blocking: true)` are currently implemented on top of `flock(2)` with blocking semantics, i.e. when it doesn't succeed immediately, the function waits until the lock is attained. This blocks the entire thread which is undesired.

Instead, the blocking semantics can be implemented in a way that they are fiber aware, i.e. `flock(2)` is used in non-blocking mode and if it would block, we yield the fiber and try again next time. This repeats until the lock is attained and opens an opportunity for other fibers to execute in the meantime.

A basic implementation could look like this:

```cr
def flock_exclusive(file)
  op = LibC::FlockOp::EX | LibC::FlockOp::NB

  while true
    if 0 == LibC.flock(file.fd, op)
      yield
    else
      errno = Errno.value
      if errno == Errno::EWOULDBLOCK
        Fiber.yield
        next
      else
        raise IO::Error.from_os_error("Error applying or removing file lock", errno)
      end
    end
  end
end
```

This is very similar to the (non-blocking) [implementation of `getrandom`](https://github.com/crystal-lang/crystal/blob/d5e6362174260282c7a16df7dd46a23a8ac8c0ab/src/crystal/system/unix/getrandom.cr#L102-L114).

### PR DESCRIPTION
**Pull request by straight-shoota**
Resolves #12721

### CODE REVIEW

File: spec/std/file_spec.cr [hunk 1]
```diff
@@ -958,6 +958,40 @@ describe "File" do
         end
       end
     end
+
+    it "#flock_shared non-blocking" do
```
Comment (beta-ziliani) [inline spec/std/file_spec.cr:962]: It's kinda confusing that this is when `blocking = true`, but it's "non-blocking". Also, this is not yet implemented in win32.

Perhaps
```suggestion
    pending_win32 "#flock_shared yielding when blocking" do
```

File: spec/std/file_spec.cr [hunk 2]
```diff
@@ -958,6 +958,40 @@ describe "File" do
...
+            done.send nil
+          end
+
+          file2.flock_shared
+          done.receive
+        end
+      end
+    end
+
+    it "#flock_exclusive non-blocking" do
```
Comment (beta-ziliani) [inline spec/std/file_spec.cr:979]: ditto
```suggestion
    pending_win32 "#flock_exclusive yielding when blocking" do
```

**Review by beta-ziliani — CHANGES_REQUESTED** [review]

**Comment by beta-ziliani** [comment]
(changes are requested for the `pending_win32` part, the name of the test is open to debate!)

**Review by beta-ziliani — APPROVED** [review]

File: src/crystal/system/unix/file.cr [hunk 3]
```diff
@@ -252,14 +252,29 @@ module Crystal::System::File
...
-  private def flock(op : LibC::FlockOp, blocking : Bool = true)
-    op |= LibC::FlockOp::NB unless blocking
+  private def flock(op : LibC::FlockOp, retry : Bool) : Nil
+    op |= LibC::FlockOp::NB
 
-    if LibC.flock(fd, op) != 0
-      raise IO::Error.from_errno("Error applying or removing file lock")
+    if retry
+      until flock(op)
+        ::Fiber.yield
```
Comment (russdill) [inline src/crystal/system/unix/file.cr:260]: If no other fibers have work to be done, what prevents this from spinning and using up 100% of the CPU until the file lock is obtained?

Comment (beta-ziliani) [reply src/crystal/system/unix/file.cr:260]: Fair point. A `sleep 5.milliseconds` will significantly lower the pressure on the CPU in such scenario without blocking for too long.

Comment (straight-shoota) [reply src/crystal/system/unix/file.cr:260]: Yeah that's definitely not ideal. This is just a basic first step and we're missing other features like a timeout (as suggested in https://github.com/crystal-lang/crystal/issues/12721#issuecomment-1308276097). Timeout and sleep are related, so might make sense to consider that together.

Comment (vgramkris) [reply src/crystal/system/unix/file.cr:260]: Do we know when the timeout feature will be available in the crystal release for this function?

Comment (straight-shoota) [reply src/crystal/system/unix/file.cr:260]: Actually, I'm not sure about the timeout. Maybe that should rather be handled more generically.

We should definitely sleep the fiber, though. I've posted a patch for that in #12861.

**Pull request merged**